### PR TITLE
Foreshadow object serialization and deserialization plus a simple python script wrapper

### PR DESCRIPTION
**Pull request by JingJZ160**
### Description
- Foreshadow object (de)serialization including the data preparers and the estimator. Note that for the estimator, if a user uses a pipeline rather than a single estimator, we cannot guarantee there is no verbose output in the serialized file.
- Included a simple oneliner script to run foreshadow from command line: `./bin/foreshadow_runner.py <data> <target>`

### CODE REVIEW

File: foreshadow/console.py [hunk 1]
```diff
@@ -1,269 +1,266 @@
...
+                "estimator from sklearn.linear_model".format(arg)
+            )
+    else:
+        return (
+            LinearRegression()
+            if determine_problem_type(y_train) == "regression"
+            else LogisticRegression()
+        )
+
+
```
Comment (cchoquette) [inline foreshadow/console.py]: This is just an uncomment, right?

File: foreshadow/foreshadow.py [hunk 2]
```diff
@@ -311,6 +316,110 @@ def score(self, data_df, y_df=None, sample_weight=None):
...
+        ret_tf = cls(**params)
+        ret_tf.data_columns = data_columns
+        if is_meta_estimator:
+            ret_tf.estimator = MetaEstimator(
+                estimator=estimator, preprocessor=ret_tf.y_preparer
+            )
+        return ret_tf
+
+    @classmethod
+    def __reconstruct_estimator(cls, data, is_meta_estimator):
```
Comment (cchoquette) [inline foreshadow/foreshadow.py]: single underscore

File: foreshadow/foreshadow.py [hunk 3]
```diff
@@ -311,6 +316,110 @@ def score(self, data_df, y_df=None, sample_weight=None):
...
+        # TODO untested on non-meta-estimator
+        if is_meta_estimator:
```
Comment (cchoquette) [inline foreshadow/foreshadow.py]: why not just make this the default and not use this flag?

File: foreshadow/utils/common.py [hunk 4]
```diff
@@ -106,3 +107,27 @@ def configure_column_sharer(self, column_sharer):
...
+
+
+class CustomizeParamsMixin:
+    """Mixin that customizes params returned from get_params."""
+
+    def customize_serialization_params(self, params):
+        """Select params that are only necessary.
+
+        Args:
+            params: the params returned from get_params
```
Comment (cchoquette) [inline foreshadow/utils/common.py]: why not make this method call get_params and force that this method only works with objects that have a working get_params method?

File: foreshadow/utils/common.py [hunk 5]
```diff
@@ -106,3 +107,27 @@ def configure_column_sharer(self, column_sharer):
         """
         if hasattr(self, "column_sharer"):
             self.column_sharer = column_sharer
+
+
+class CustomizeParamsMixin:
+    """Mixin that customizes params returned from get_params."""
+
+    def customize_serialization_params(self, params):
+        """Select params that are only necessary.
```
Comment (cchoquette) [inline foreshadow/utils/common.py]: Update this doc string. I believe it should say: "Select subset of get_params on this class' init."
Two major questions:
1. (for my own knowledge). Why was this needed instead of just using get_params?
2. Shouldn't "self" and "kwargs" already not be in the params since they were returned from self.get_params()?
In which case, see my comment below about the for loop and how it could be augmented.

File: foreshadow/utils/common.py [hunk 6]
```diff
@@ -106,3 +107,27 @@ def configure_column_sharer(self, column_sharer):
...
+            params: the params returned from get_params
+
+        Returns:
+            selected_params: params that are selected.
+
+        """
+        import inspect
+
+        init_params = inspect.signature(self.__init__).parameters
+        selected_params = {
```
Comment (cchoquette) [inline foreshadow/utils/common.py]: See above docstring comment, this could be:
`selected_params = {name: params.get('name', None) for name in init_params if name in params}`
Also, .pop removes the key from the dict and .get doesnt. I would prefer to use .get unless you explicitly want the former action to prevent future maintainers from running into weird bugs because they miss that .pop() was used instead of .get().

File: foreshadow/console.py (hunk 1 above)
Comment (JingJZ160) [reply foreshadow/console.py]: Mostly, yes. The only additional change is that I added a line of code to serialize the whole foreshadow object instead of the X_preprocessor and y_preprocessor in the old code.

File: foreshadow/estimators/meta.py [hunk 7]
```diff
@@ -19,6 +22,32 @@ def __init__(self, estimator, preprocessor):
...
 
+    def dict_serialize(self, deep=True):  # noqa
+        params = self.get_params(deep)
+        selected_params = self.__create_selected_params(params)
+        serialized = _make_serializable(
+            selected_params, serialize_args=self.serialize_params
+        )
+        return serialized
+
+    def __create_selected_params(self, params):
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/estimators/meta.py]: Good catch. It should be in the CustomizeParamsMixin.

File: foreshadow/utils/common.py [hunk 8]
```diff
@@ -106,3 +107,27 @@ def configure_column_sharer(self, column_sharer):
...
+            name: params.pop(name)
+            for name in init_params
+            if name not in ["self", "kwargs"]
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/utils/common.py]: On a second thought I'm not sure if we need this Mixin. I may be able to achieve the same effect for the classes that inherits this Mixin by just call `.dict_serialize(deep=False)`.

File: foreshadow/foreshadow.py [hunk 9]
```diff
@@ -311,6 +316,110 @@ def score(self, data_df, y_df=None, sample_weight=None):
...
+
+        serialized_estimator = estimator.get_params()
+        serialized_estimator["_class"] = (
+            estimator.__module__ + "." + type(estimator).__name__
+        )
+        serialized_estimator["_method"] = "dict"
+
+        if is_meta_estimator:
+            result["estimator"] = serialized_estimator
+            result["preprocessor"] = "_y_preparer"
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/foreshadow.py]: You are correct. This if statement is unnecessary. The only caveat is that when we deserialize the foreshadow object back, the estimator is not a MetaEstimator unless we call `.fit` on the foreshadow object. I'm considering moving the following lines to the `__init__` method of the Foreshadow class:

```
# setup MetaEstimator if y_preparer is passed in
if self.y_preparer is not None:
    self.estimator = MetaEstimator(self.estimator, self.y_preparer)
```

File: foreshadow/foreshadow.py [hunk 10]
```diff
@@ -311,6 +316,110 @@ def score(self, data_df, y_df=None, sample_weight=None):
...
+        serialized = _make_serializable(
+            selected_params, serialize_args=self.serialize_params
+        )
+        serialized["estimator"] = self.__customize_serialized_estimator(
+            self.estimator
+        )
+        return serialized
+
+    @staticmethod
+    def __customize_serialized_estimator(estimator):
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/foreshadow.py]: will do

File: foreshadow/foreshadow.py (hunk 2 above)
Comment (JingJZ160) [reply foreshadow/foreshadow.py]: will do

File: foreshadow/foreshadow.py (hunk 3 above)
Comment (JingJZ160) [reply foreshadow/foreshadow.py]: This argument won't be necessary in the latest change.

File: foreshadow/utils/common.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/utils/common.py]: That makes sense and should work!

File: foreshadow/foreshadow.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (cchoquette) [reply foreshadow/foreshadow.py]: This might be the best time to clean up the properties that were used. I don't think the properties is the best way of auto initializing those attributes and think they should all be moved to the init, similar to what you just proposed :)
